### ISSUE TITLE
Advice on transfomer_big hyperparameters?

### ISSUE THREAD

**Issue by mzeidhassan** [modernmt/modernmt#444]
Hi MMT team,

I want ask you about the best hyper-parameters to use with transformer_big, based on your own experience.

According to [this paper](https://arxiv.org/abs/1804.00247), the author suggests that we should use 'transformer_big' with higher batch size.

I am testing an engine, and it had been running for a few days, but I see this in tensorboard.
![image](https://user-images.githubusercontent.com/7589948/52495749-e3e07800-2b8e-11e9-952c-a0a762b05e16.png)

The command used is:

`-e en-fr-transbig \
        --train-steps=570000 \
        --hparams=transformer_big \
        --batch-size=2048
`
Do you have any idea what could be causing this? Is there is something wrong with the command? 
Have you tried using tranformer_big at your end? Is the quality better than transformer_base? 

I appreciate your advice. Thanks in advance!

**Comment by nicolabertoldi** [comment]
Actually we followed the suggestions by Poepn and Bojar.
We normally use transformer_big, because it definitely outperforms transformer_base, especially when the training data is reasonably large.

In particular, we use:
- a larger batch_size of 3072, which usually fit the GPU RAM (11Gb)
- a learning_rate_warmup_steps of 8000

**Comment by mzeidhassan** [comment]
Thanks @nicolabertoldi for the advice and for getting back to me on this.

1) May I ask you please what exact command did you use to start transformer_big? 

2) How long did it take to converge? 
I kept it running for a couple of days, and eventually I had to stop it because it didn't converge for some reason. 

3) When you say a reasonably large data, how many bilingual pairs are we talking about here?

Thanks again for your support.

**Comment by nicolabertoldi** [comment]
1. To activate transofrmer big you can run this command
```mmt create``` 
with this additional parameter 
```--hparams transformer_big```

2. Convergence depends on many factors: amount of training data, batch_size, complexity of the domain, and so on. It is very hard to establish a golden rule for deciding when the stop the training.
This is our empirical but consolidated policy:
i. every 10k steps we keep the latest checkpoint
ii. we evaluate the performance of this checkpoint on an independent test set
iii. we stop the training if we observe a sort of plateau of the performance

3.The concept of "reasonably large" cannot be defined well, because it strongly depends on the complexity of the languages, on the difficulty and generalness of the application domain, and so on.
We can suggest to use as much data as possible, provided that data are related with your application domain and are not too noisy.

**Comment by mzeidhassan** [comment]
Thanks @nicolabertoldi  for your follow-up. I did use the correct hparam as you explained above,  but in my case, it was running and as you can see in the screen shot above, there was no real convergence even after 4 days of training. So, I am not sure what was wrong. 

I will try with a batch_size of 3072 and see. I also assume that "learning_rate_warmup_steps" can be passed as a hyper-parameter "--learning_rate_warmup_steps 8000" in the create command, right? Or do I need to change it somewhere in the code?

Thanks again for your support!

**Comment by nicolabertoldi** [comment]
@mzeidhassan 
Currently, we expose a limited set of all tensor2tensor parameters. So at the moment it is not possible to set parameter `learning_rate_warmup_steps` in the command line. Probably, we add this feature in next versions.

However, to get what you want, you can simply modify one line of the source code in this file
`cli/mmt/engine.py`.

You just need to replace line 172
```hparams_p = 'batch_size=%d' % batch_size```
with the following
```hparams_p = 'batch_size=%d,learning_rate_warmup_steps=8000' % batch_size```

Similarly, you can modify most of the T2T training parameters.

**Comment by mzeidhassan** [comment]
Thank you @nicolabertoldi so much for your support. I appreciate it. 
I will test it.

**Comment by davidecaroselli** [comment]
I'm closing this for inactivity but please feel free to re-open it for further discussions!
